examples/mnist/Main.hs

```diff
@@ -0,0 +1,159 @@
+  deriving (Show, Eq)
+
+data Linear (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) =
+  Linear { weight :: Parameter dtype '[inputFeatures, outputFeatures]
+         , bias :: Parameter dtype '[outputFeatures]
+         } deriving (Show, Generic)
+
+linear
+  :: Linear dtype inputFeatures outputFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+linear Linear {..} input =
+  add (mm input (toDependent weight)) (toDependent bias)
+
+makeIndependent :: Tensor dtype shape -> IO (Parameter dtype shape)
+makeIndependent t = Parameter <$> A.makeIndependent (toDynamic t)
+
+instance A.Parameterized (Linear dtype inputFeatures outputFeatures)
+
+instance (KnownDType dtype, KnownNat inputFeatures, KnownNat outputFeatures) => A.Randomizable (LinearSpec dtype inputFeatures outputFeatures) (Linear dtype inputFeatures outputFeatures) where
+  sample LinearSpec =
+    Linear <$> (makeIndependent =<< randn) <*> (makeIndependent =<< randn)
+
+data MLPSpec (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) (hiddenFeatures :: Nat) = MLPSpec
+
+data MLP (dtype :: D.DType) (inputFeatures :: Nat) (outputFeatures :: Nat) (hiddenFeatures :: Nat) =
+  MLP { layer0 :: Linear dtype inputFeatures hiddenFeatures
+      , layer1 :: Linear dtype hiddenFeatures hiddenFeatures
+      , layer2 :: Linear dtype hiddenFeatures outputFeatures
+      } deriving (Show, Generic)
+
+instance A.Parameterized (MLP dtype inputFeatures outputFeatures hiddenFeatures)
+
+instance (KnownDType dtype, KnownNat inputFeatures, KnownNat outputFeatures, KnownNat hiddenFeatures) => A.Randomizable (MLPSpec dtype inputFeatures outputFeatures hiddenFeatures) (MLP dtype inputFeatures outputFeatures hiddenFeatures) where
+  sample MLPSpec =
+    MLP <$> A.sample LinearSpec <*> A.sample LinearSpec <*> A.sample LinearSpec
+
+mlp
+  :: MLP dtype inputFeatures outputFeatures hiddenFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+mlp MLP {..} = linear layer2 . relu . linear layer1 . relu . linear layer0
+
+model
+  :: MLP dtype inputFeatures outputFeatures hiddenFeatures
+  -> Tensor dtype '[batchSize, inputFeatures]
+  -> Tensor dtype '[batchSize, outputFeatures]
+model = (sigmoid .) . mlp
+
+foldLoop
+  :: forall a b m . (Num a, Enum a, Monad m) => b -> a -> (b -> a -> m b) -> m b
+foldLoop x count block = foldM block x ([1 .. count] :: [a])
+
+type BatchSize = 64
+type TestSize = 10000
+
+
+randomIndexes :: [Int]
+randomIndexes = map (`mod` natValI @I.DataDim) $ randoms seed
+  where
+    seed = mkStdGen 123
+
+
+toBackend :: String -> Tensor dtype shape -> Tensor dtype shape
+toBackend backend t =
+  case backend of
+    "CUDA" -> toCUDA t
+    _ -> toCPU t
+
+
+
+main = do
+  backend' <- try (getEnv "BACKEND") :: IO (Either SomeException String)
+  let backend = 
+        case backend' of
+          Right "CUDA" -> "CUDA"
+          _ -> "CPU"
```

Comment:
In Pytorch, the best way to deal with this is through devices rather than backends. That lets you deal with multi-cpu and multi-gpu sitatuations.